elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/query/DefaultQueryPlanMerger.java

```diff
@@ -0,0 +1,146 @@
+ */
+public class DefaultQueryPlanMerger implements QueryPlanMerger {
+    MetaDataStore metaDataStore;
+
+    public DefaultQueryPlanMerger(MetaDataStore metaDataStore) {
+        this.metaDataStore = metaDataStore;
+    }
+
+    @Override
+    public boolean canMerge(QueryPlan a, QueryPlan b) {
+        Preconditions.checkNotNull(a);
+        Preconditions.checkNotNull(b);
+
+        if (a.nestDepth() != b.nestDepth()) {
+            if (a.nestDepth() > b.nestDepth() && !b.canNest(metaDataStore)) {
+                return false;
+            }
+
+            if (b.nestDepth() > a.nestDepth() && !a.canNest(metaDataStore)) {
+                return false;
+            }
+
+            a = nestToDepth(a, b.nestDepth());
+            b = nestToDepth(b, a.nestDepth());
+        }
+
+        boolean result = canMergeFilter(a, b);
+        if (! result) {
+            return false;
+        }
+
+        /*
+         * Metrics can always coexist provided they have different aliases (which is enforced by the API).
+         */
+
+        result = canMergeDimensions(a, b);
+        if (! result) {
+            return false;
+        }
+
+        result = canMergeTimeDimensions(a, b);
+
+        if (!result) {
+            return false;
+        }
+
+        if (a.isNested()) {
+            result = canMerge((QueryPlan) a.getSource(), (QueryPlan) b.getSource());
+        }
+
+        return result;
+    }
+
+    protected boolean canMergeDimensions(QueryPlan a, QueryPlan b) {
+        for (DimensionProjection dimension : a.getDimensionProjections()) {
+            DimensionProjection otherDimension = b.getDimensionProjection(dimension.getName());
+
+            if (otherDimension != null && ! Objects.equals(otherDimension.getArguments(), dimension.getArguments())) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    protected boolean canMergeTimeDimensions(QueryPlan a, QueryPlan b) {
+        for (TimeDimensionProjection dimension : a.getTimeDimensionProjections()) {
+            TimeDimensionProjection otherDimension = b.getTimeDimensionProjection(dimension.getName());
+
+            if (otherDimension != null && ! Objects.equals(otherDimension.getArguments(), dimension.getArguments())) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    protected boolean canMergeFilter(QueryPlan a, QueryPlan b) {
+        if (! Objects.equals(a.getWhereFilter(), b.getWhereFilter())) {
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public QueryPlan merge(QueryPlan a, QueryPlan b) {
+        Preconditions.checkNotNull(a);
+        Preconditions.checkNotNull(b);
+
+        a = nestToDepth(a, b.nestDepth());
+        b = nestToDepth(b, a.nestDepth());
+
+        assert (a.isNested() || a.getSource().equals(b.getSource()));
+
+        Set<MetricProjection> metrics = Streams.concat(b.getMetricProjections().stream(),
```

Comment:
break out into 'protected mergetFoo(QP a, QP b)` for each of the terms, including where Filters/